core/state/statedb.go

```diff
@@ -174,6 +174,8 @@ type StateDB struct {
 
 	// Testing hooks
 	onCommit func(states *triestate.Set) // Hook invoked when commit is performed
+
+	collectWitness bool
```

Comment:
Wondering if this should be part a field on the prefertcher itself. It would make the statedb less complicated if it didn't have yet another thing to worry about. Then we could always blindly call prefetcher.schedule and have that early return if it doesn't want to prefetch.

Reply:
> Wondering if this should be part a field on the prefertcher itself. It would make the statedb less complicated if it didn't have yet another thing to worry about.

Well, when we move to collecting code/block-hashes, we will need  an API for the EVM to determine whether witnesses are being collected.  In the PR that implements full stateless collection/execution, I've added a `Witness` method to the `StateDB` which returns a non-nil witness when collection is enabled.

> Then we could always blindly call prefetcher.schedule and have that early return if it doesn't want to prefetch.

I don't follow what you mean by "always blindly call prefetcher.schedule".  If we move more of the logic into the prefetcher, we could add an extra prefetcher method `prefetchForWitness` which is called when we are prefetching for read values (and ends up being a noop if the witness collection isn't enabled).  But I'm not sure this is what you mean.